### PR TITLE
pcp: init at 0.4.0

### PR DESCRIPTION
**Pull request by MatthewCroughan**
###### Motivation for this change

Adds `pcp` which is a terminal file sharing program much like magic-wormhole and croc.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by zowoq — COMMENTED** [review]
commit should be `pcp: init at 0.4.0`

File: pkgs/tools/misc/pcp/default.nix [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  src = fetchFromGitHub {
+    owner = "dennis-tra";
+    repo = "pcp";
+    rev = "v${version}";
+    sha256 = "sha256-aZO8VuOiYhOPctFKZ6a2psJB0lKHlPc+NLy2RWDU4JI=";
+  };
+
+  vendorSha256 = "sha256-fXDOOxkl4Rws28uUmOO4MxljrJPIXgWo9YYkKazr18A=";
+
+  runVend = true;
```
Comment (zowoq) [inline pkgs/tools/misc/pcp/default.nix:16]: This doesn't seem to be needed on darwin or linux?

**Comment by MatthewCroughan** [comment]
All feedback so far has been resolved.

File: pkgs/tools/misc/pcp/default.nix [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  version = "0.4.0";
+
+  src = fetchFromGitHub {
+    owner = "dennis-tra";
+    repo = "pcp";
+    rev = "v${version}";
+    sha256 = "sha256-aZO8VuOiYhOPctFKZ6a2psJB0lKHlPc+NLy2RWDU4JI=";
+  };
+
+  vendorSha256 = "sha256-fXDOOxkl4Rws28uUmOO4MxljrJPIXgWo9YYkKazr18A=";
```
Comment (zowoq) [inline pkgs/tools/misc/pcp/default.nix:14]: Needs to be updated now `runVend` is removed.

```suggestion
  vendorSha256 = "sha256-3bkzBQ950Phg4A9p+IjeUx7Xw7eVmUbeYnQViNjghFk=";
```

Comment (MatthewCroughan) [reply pkgs/tools/misc/pcp/default.nix:14]: You cannot call `buildGoModule` without the required `vendorSha256` argument:
```
user: matthew swordfish in nixpkgs on  add-pcp [$!] via ❄️  impure (shell) took 3s 
❯ nix build .#pcp -L
warning: Git tree '/home/matthew/git/nixpkgs' is dirty
error: anonymous function at /nix/store/fsa9rjw3ni69xv10wassi686r7pgcrfz-source/pkgs/development/go-modules/generic/default.nix:3:1 called without required argument 'vendorSha256'

       at /nix/store/fsa9rjw3ni69xv10wassi686r7pgcrfz-source/pkgs/tools/misc/pcp/default.nix:3:1:

            2|
            3| buildGoModule rec {
             | ^
            4|   pname = "pcp";
```

Comment (MatthewCroughan) [reply pkgs/tools/misc/pcp/default.nix:14]: My bad, I thought you were telling me to get rid of the argument on first glance.

**Comment by MatthewCroughan** [comment]
@legendofmiracles I hastily fixed your suggestion without realising you had made a suggestion via Github. I have added your Co-author string, since I did not realise and failed to click the "commit suggestion" button. Thank you.

**Comment by zowoq** [comment]
Tests fail on darwin ofborg but they passed for me locally.

Probably a sandboxing issue, `__darwinAllowLocalNetworking` may fix it (may still fail on hydra anyway) but I'm fine with just setting `doCheck = !stdenv.isDarwin`.

**Comment by Kranzes** [comment]
> commit should be `pcp: init at 0.4.0`

Same thing for the PR title.

**Comment by bobby285271** [comment]
[Test](https://github.com/dennis-tra/pcp/blob/c7a005be81126a26c96c13409d3dfaba73f00ae4/pkg/dht/discoverer_test.go#L186-L187) fails [on darwin](https://logs.nix.ci/?key=nixos/nixpkgs.152869&attempt_id=39d48700-01b2-4dd7-bbea-e73c08a117cf), shall we disable it and/or report it to upstream?

**Comment by onny** [comment]
Result of `nixpkgs-review pr 152869` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>pcp</li>
  </ul>
</details>

**Comment by onny** [comment]
Result of `nixpkgs-review pr 152869` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>pcp</li>
  </ul>
</details>

**Comment by onny** [comment]
Tests fail on my system
```
=== RUN   TestTimeCriticalProtocol_Bootstrap_connectsBootstrapPeersInParallel
    protocol_test.go:127: 
                Error Trace:    protocol_test.go:127
                Error:          Max difference between 50ms and 76.904548ms allowed is 2.5e+07, but difference was -2.6904548e+07
                Test:           TestTimeCriticalProtocol_Bootstrap_connectsBootstrapPeersInParallel
--- FAIL: TestTimeCriticalProtocol_Bootstrap_connectsBootstrapPeersInParallel (0.17s)
=== RUN   TestProtocol_Bootstrap_errorOnNoConfiguredBootstrapPeers
--- PASS: TestProtocol_Bootstrap_errorOnNoConfiguredBootstrapPeers (0.00s)
=== RUN   TestProtocol_Bootstrap_cantConnectToOneLessThanThreshold
--- PASS: TestProtocol_Bootstrap_cantConnectToOneLessThanThreshold (0.00s)
=== RUN   TestProtocol_Bootstrap_cantConnectToMultipleLessThanThreshold
--- PASS: TestProtocol_Bootstrap_cantConnectToMultipleLessThanThreshold (0.00s)
=== RUN   TestProtocol_Bootstrap_cantConnectButGreaterThanThreshold
--- PASS: TestProtocol_Bootstrap_cantConnectButGreaterThanThreshold (0.00s)
=== RUN   TestProtocol_DiscoveryIdentifier_returnsCorrect
--- PASS: TestProtocol_DiscoveryIdentifier_returnsCorrect (0.00s)
FAIL
FAIL    github.com/dennis-tra/pcp/pkg/dht       0.634s
FAIL
```

**Comment by MatthewCroughan** [comment]
@onny I underclocked my laptop to 500mhz and ran `nix build`, and indeed the tests fail. If you're running the tests on too slow a machine, they will fail. The clue is in the name of the test `TestTimeCriticalProtocol`. It is "time critical", which means it will fail if the test does not complete fast enough. Not sure what to do to fix this, other than disable the checks entirely.

**Comment by onny** [comment]
> @onny I underclocked my laptop to 500mhz and ran `nix build`, and indeed the tests fail. If you're running the tests on too slow a machine, they will fail. The clue is in the name of the test `TestTimeCriticalProtocol`. It is "time critical", which means it will fail if the test does not complete fast enough. Not sure what to do to fix this, other than disable the checks entirely.

Oh wow thats interesting. We should disable this specific test if possible

**Comment by onny** [comment]
Result of `nixpkgs-review pr 152869` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>pcp</li>
  </ul>
</details>

**Comment by onny** [comment]
Thank you!

**Pull request merged**